<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="" xml:lang="">
<head>
  <meta charset="utf-8" />
  <meta name="generator" content="pandoc" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes" />
  <title>parts/4.7.ErrorHandling.md</title>
  <style>
    html {
      line-height: 1.5;
      font-family: Georgia, serif;
      font-size: 20px;
      color: #1a1a1a;
      background-color: #fdfdfd;
    }
    body {
      margin: 0 auto;
      max-width: 36em;
      padding-left: 50px;
      padding-right: 50px;
      padding-top: 50px;
      padding-bottom: 50px;
      hyphens: auto;
      word-wrap: break-word;
      text-rendering: optimizeLegibility;
      font-kerning: normal;
    }
    @media (max-width: 600px) {
      body {
        font-size: 0.9em;
        padding: 1em;
      }
    }
    @media print {
      body {
        background-color: transparent;
        color: black;
        font-size: 12pt;
      }
      p, h2, h3 {
        orphans: 3;
        widows: 3;
      }
      h2, h3, h4 {
        page-break-after: avoid;
      }
    }
    p {
      margin: 1em 0;
    }
    a {
      color: #1a1a1a;
    }
    a:visited {
      color: #1a1a1a;
    }
    img {
      max-width: 100%;
    }
    h1, h2, h3, h4, h5, h6 {
      margin-top: 1.4em;
    }
    h5, h6 {
      font-size: 1em;
      font-style: italic;
    }
    h6 {
      font-weight: normal;
    }
    ol, ul {
      padding-left: 1.7em;
      margin-top: 1em;
    }
    li > ol, li > ul {
      margin-top: 0;
    }
    blockquote {
      margin: 1em 0 1em 1.7em;
      padding-left: 1em;
      border-left: 2px solid #e6e6e6;
      color: #606060;
    }
    code {
      font-family: Menlo, Monaco, 'Lucida Console', Consolas, monospace;
      font-size: 85%;
      margin: 0;
    }
    pre {
      margin: 1em 0;
      overflow: auto;
    }
    pre code {
      padding: 0;
      overflow: visible;
    }
    .sourceCode {
     background-color: transparent;
     overflow: visible;
    }
    hr {
      background-color: #1a1a1a;
      border: none;
      height: 1px;
      margin: 1em 0;
    }
    table {
      margin: 1em 0;
      border-collapse: collapse;
      width: 100%;
      overflow-x: auto;
      display: block;
      font-variant-numeric: lining-nums tabular-nums;
    }
    table caption {
      margin-bottom: 0.75em;
    }
    tbody {
      margin-top: 0.5em;
      border-top: 1px solid #1a1a1a;
      border-bottom: 1px solid #1a1a1a;
    }
    th {
      border-top: 1px solid #1a1a1a;
      padding: 0.25em 0.5em 0.25em 0.5em;
    }
    td {
      padding: 0.125em 0.5em 0.25em 0.5em;
    }
    header {
      margin-bottom: 4em;
      text-align: center;
    }
    #TOC li {
      list-style: none;
    }
    #TOC a:not(:hover) {
      text-decoration: none;
    }
    code{white-space: pre-wrap;}
    span.smallcaps{font-variant: small-caps;}
    span.underline{text-decoration: underline;}
    div.column{display: inline-block; vertical-align: top; width: 50%;}
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
    ul.task-list{list-style: none;}
    .display.math{display: block; text-align: center; margin: 0.5rem auto;}
  </style>
  <!--[if lt IE 9]>
    <script src="//cdnjs.cloudflare.com/ajax/libs/html5shiv/3.7.3/html5shiv-printshiv.min.js"></script>
  <![endif]-->
</head>
<body>
<header id="title-block-header">
<h1 class="title">parts/4.7.ErrorHandling.md</h1>
</header>
<h2 id="error-handling">4.7 Error Handling</h2>
<p>The C++ language and standard library provides several mechanisms for error handling:</p>
<ul>
<li>exceptions, signaling a contract violation</li>
<li>returning just an error code with a special value denoting success.</li>
<li>returning a special error value of the function’s return type, such as <code>std::string::npos</code></li>
<li>combining the return value with an error indication, e.g., through <code>std::expected&lt;T,E&gt;</code> or <code>std::optional&lt;T&gt;</code> for a return type <code>T</code>. The error type <code>E</code> can use <code>std::error_code</code>, for example.</li>
<li>marking the error status as a side effect, such as with the “global” variable <code>errno</code>, or as an objects status as with <code>std::istream</code>’s flags.</li>
</ul>
<p>Be aware, that many parts of the standard library specify preconditions and undefined behavior results if those aren’t met, for example, <code>std::vector&lt;int&gt;::front()</code> requires a non-empty vector, when called.</p>
<p>Exceptions allow for errors to be propagated up the call chain. Throwing an exception due to a detected error situation allows the error to be handled at an appropriate level in a corresponding catch block. As the exception propagates to its handler, local objects are destroyed appropriately in reverse order of their construction; this mechanism is known as <em>stack unwinding</em>. A search for a matching handler stops at</p>
<ul>
<li>the function <code>main()</code>,</li>
<li>a function declared <code>noexcept</code>, and</li>
<li>the function of a thread.</li>
</ul>
<p>An exception propagated from constructors of non-local variables and destructors of variables with static storage duration can never have a matching handler.</p>
<p>Failing to provide a matching handler on the call chain for an exception thrown causes a call to <code>std::terminate()</code> and the program terminates.</p>
<p>When calling non-returning program termination functions like <code>abort()</code>, <code>std::terminate()</code>, or <code>exit()</code>, the program terminates without <em>stack unwinding</em>.</p>
<p>#TODO:</p>
<p>For coroutines, a rethrown exception or an uncaught exception causes the coroutine to end and cannot be resume. The exception itself is not propagated.</p>
<ul>
<li><pre><code> (Ensure that vulnerability associated with this mechanism is documented in 6.36)       </code></pre></li>
<li><pre><code>  Distinct mechanism for IEEE floating point exceptions</code></pre></li>
<li><pre><code>     defer to C</code></pre></li>
<li>assertions</li>
<li><pre><code>  terminates program and all threads and tasks if assertion failure happens in main thread?</code></pre></li>
<li><pre><code>  terminates thread if raised in a thread?</code></pre></li>
<li><pre><code>  terminates task (and more?) if raised in a task.</code></pre></li>
<li>[basic.start.name] para 4</li>
<li><pre><code> terminating program without leaving current block (say calling exit) does not invoke any destructors.</code></pre></li>
<li><pre><code>    if exit is called, during object destruction with static or thread storage duration, program has undefined behaviour</code></pre></li>
<li>errno values and std::error_code</li>
<li><pre><code>  Used by any library functions external to C++, and by some C++ libraries.</code></pre></li>
<li><pre><code>  errno is a thread-local single value that, if non-zero, reports an error that has occurred and is returned by function as the return value.</code></pre></li>
<li><pre><code>     Any external can be called using error_code which will wrap the errno.</code></pre></li>
<li><pre><code>  error_code is an extension or errno but is not a global variable and can work with classes for better reporting.</code></pre></li>
<li>signals</li>
<li><pre><code> operating system dependant</code></pre></li>
<li><pre><code> see what 24772-3 C says</code></pre></li>
<li>system_error</li>
<li>stacktrace</li>
<li>std::expected</li>
<li></li>
<li>For 6.36</li>
<li><pre><code> Vulnerability in mixing of different error mechanisms</code></pre></li>
</ul>
<h2 id="also-two-scenarios">Also two scenarios</h2>
<p>In addition to that list, an operating system might use the “signal” mechanism to notify a running C++ program. A signal-handler can be defined to act asynchronously upon a sent signal that isn’t ignored.</p>
</body>
</html>
